### PR TITLE
[Feature] Adding callback functionalities in train functions

### PR DESCRIPTION
**Pull request by chMoussa**
Related to a comment from #202, we can use callback functions in train functions to extend possibilities to do more during these functions and allow flexibility. Similar to scipy.optimize, we have an `OptimizeResult` class storing several optimization-related object/values.  The callback should take an `OptimizeResult` as input and apply a user-defined function.

Several booleans locate filter where the callbacks should be called within train. This also allow avoiding a few if statements for some already-defined callback-type functions (writing, checkpointing, printing, etc).

### CODE REVIEW

File: qadence/ml_tools/config.py [hunk 1]
```diff
@@ -28,6 +29,80 @@
...
+        self.call_before_opt = call_before_opt
+        self.call_end_epoch = call_end_epoch
+        self.call_after_opt = call_after_opt
+        self.call_during_eval = call_during_eval
+
+        self.every = every
+
+        if callback_condition is None:
+            if self.every > 1:
+                self.callback_condition = lambda opt_result: opt_result.iteration % self.every == 0
```
Comment (gvelikova) [inline qadence/ml_tools/config.py:95]: This will always be true for self.every==1, but I guess running it each time unnecessary would be costing extra.

**Review by gvelikova — APPROVED** [review]

File: qadence/ml_tools/config.py [hunk 2]
```diff
@@ -28,6 +29,80 @@
...
+            else:
+                self.callback_condition = lambda opt_result: True
+        else:
+            self.callback_condition = callback_condition
+
+    def __call__(self, opt_result: OptimizeResult, *args: Any, **kwargs: Any) -> Any:
+        if self.callback_condition(opt_result):
+            return self.callback(opt_result, *args, **kwargs)
+
```
Comment (debrevitatevitae) [inline qadence/ml_tools/config.py:104]: As a general comment here, I would delegate the responsibility of triggering the callback condition to the callback condition itself. Here is a possible option
```python
@dataclass
class TrainingStep:
    """Information about which phase of the training we are in."""
    epoch_number: int | None = None
    before_opt: bool = True
    after_opt: bool = False
    during_eval: bool = False

CallbackFunction = Callable[[OptimizeResults], None]
CallbackCondition = Callable[[TrainingStep], bool]


class Callback:
    def __init__(self, callback_function: CallbackFunction, callback_condition: CallbackCondition):
        self._callback_function = callback_function
        self._callback_condition = callback_condition

    def __call__(self, training_step: TrainingStep, opt_result: OptimizeResult):
        if self.callback_condition(training_step):
            return self._callback_function(opt_result)
```

`TrainingStep` would be an internal 'clock' for the training phase, updated in `train_*` functions. As mentioned, we let `callback_condition` decide on whether the clock is right to trigger. In this way we could also remove `iteration` from `OptmizeResults`.

File: qadence/ml_tools/train_grad.py [hunk 3]
```diff
@@ -160,115 +160,167 @@ def loss_fn(model: torch.nn.Module, data: torch.Tensor) -> tuple[torch.Tensor, d
...
+    if not ((dataloader is None) or isinstance(dataloader, (DictDataLoader, DataLoader))):
+        raise NotImplementedError(
+            f"Unsupported dataloader type: {type(dataloader)}. "
+            "You can use e.g. `qadence.ml_tools.to_dataloader` to build a dataloader."
+        )
+
+    def next_loss_iter(dl_iter: Union[None, DataLoader, DictDataLoader]) -> Any:
+        xs = next(dl_iter) if dl_iter is not None else None
+        xs_to_device = data_to_device(xs, device=device, dtype=data_dtype)
+        return loss_fn(model, xs_to_device)
```
Comment (debrevitatevitae) [inline qadence/ml_tools/train_grad.py:172]: Maybe include a docstring for explanation.

File: qadence/ml_tools/train_grad.py [hunk 4]
```diff
@@ -160,115 +160,167 @@ def loss_fn(model: torch.nn.Module, data: torch.Tensor) -> tuple[torch.Tensor, d
...
+    callbacks_before_opt_eval = [
+        callback for callback in callbacks if callback.call_before_opt and callback.call_during_eval
+    ]
+
     with progress:
         dl_iter = iter(dataloader) if dataloader is not None else None
 
         # Initial validation evaluation
         try:
+            opt_result = OptimizeResult(init_iter, model, optimizer, None, dict())
```
Comment (debrevitatevitae) [inline qadence/ml_tools/train_grad.py:272]: can we avoid passing the last two arguments altogether here?

File: qadence/ml_tools/train_grad.py [hunk 5]
```diff
@@ -160,115 +160,167 @@ def loss_fn(model: torch.nn.Module, data: torch.Tensor) -> tuple[torch.Tensor, d
...
+                ),
+                every=config.checkpoint_every,
+                call_before_opt=True,
+                call_after_opt=True,
+                call_during_eval=True,
+            )
+        ]
+
+    def run_callbacks(callback_iterable: list[Callback], opt_res: OptimizeResult) -> None:
+        [callback(opt_res) for callback in callback_iterable]
```
Comment (debrevitatevitae) [inline qadence/ml_tools/train_grad.py:256]: why the list comprehension?

**Review by debrevitatevitae — COMMENTED** [review]
Cool work! Can we also add an example or a tutorial for the docs to showcase the callback definitions?

File: qadence/ml_tools/train_grad.py (hunk 5 above)
Comment (chMoussa) [reply qadence/ml_tools/train_grad.py]: Just liked one-liners but putting it back to a simple loop syntax

File: qadence/ml_tools/config.py (hunk 2 above)
Comment (chMoussa) [reply qadence/ml_tools/config.py:104]: Hm, you can also trigger `callback_condition` based on the loss value or a condition coming from model or optimizer, so that is why I kept also the `callback_condition` to take as input the `OptimizeResult` instance. Although, I don't demonstrate it here directly. Also, the callback would be missing when it is called which would help us filter them before getting in the loop, and avoid unnecessary calling `callback_condition`.

**Comment by chMoussa** [comment]
> Cool work! Can we also add an example or a tutorial for the docs to showcase the callback definitions?

Yes just did!

File: docs/tutorials/qml/ml_tools.md [hunk 6]
```diff
@@ -78,19 +78,25 @@ def loss_fn(model: torch.nn.Module, data: torch.Tensor) -> tuple[torch.Tensor, d
...
+custom_callback = Callback(lambda opt_res: print(opt_res.model.parameters()) if opt_res.loss < 1.0e-03 else print('Loss threshold not yet reached.'), every=10, call_end_epoch=True)
+
 config = TrainConfig(
     folder="some_path/",
     max_iter=n_epochs,
     checkpoint_every=100,
     write_every=100,
     batch_size=batch_size,
+    callbacks = [custom_callback]
 )
```
Comment (awennersteen) [inline docs/tutorials/qml/ml_tools.md:100]: It would perhaps be nice to show the aforementioned `callback_condition`?

Comment (chMoussa) [reply docs/tutorials/qml/ml_tools.md:100]: Sure, will change the example by making the if statement a `callback_condition` function!

File: qadence/ml_tools/train_grad.py [hunk 7]
```diff
@@ -160,115 +160,178 @@ def loss_fn(model: torch.nn.Module, data: torch.Tensor) -> tuple[torch.Tensor, d
...
-                        iteration,
-                        config.plotting_functions,
-                        tracking_tool=config.tracking_tool,
-                    )
                 if perform_val:
                     if iteration % config.val_every == 0:
-                        xs = next(dl_iter_val)
-                        xs_to_device = data_to_device(xs, device=device, dtype=data_dtype)
-                        val_loss, *_ = loss_fn(model, xs_to_device)
+                        val_loss, *_ = next_loss_iter(dl_iter_val)
```
Comment (awennersteen) [inline qadence/ml_tools/train_grad.py:328]: we had some bug elsewhere where 0 is a valid value for these `val_every` type arguments, see #511 

IDK if this is guarded already, or if it's worth to add one

File: qadence/ml_tools/config.py [hunk 8]
```diff
@@ -27,6 +28,80 @@
...
+    an OptimizeResult instance.
+
+    Attributes:
+        callback (CallbackFunction): Callback function accepting an
+            OptimizeResult as first argument.
+        callback_condition (CallbackConditionFunction | None, optional): Function that
+            conditions the call to callback. Defaults to None.
+        every (int, optional): Callback to be called each `every` epoch. Defaults to 1.
+            If callback_condition is None, we set
+            callback_condition to returns True when iteration % every == 0.
```
[the remark this answers is not in the document]
Comment (chMoussa) [reply qadence/ml_tools/config.py:48]: `called_every` sounds good to me

**Review by awennersteen — APPROVED** [review]
Some minor, optional, suggestions.

File: qadence/ml_tools/train_grad.py (hunk 7 above)
Comment (chMoussa) [reply qadence/ml_tools/train_grad.py:328]: I think we don't check. But I would raise an error in case it is 0 cause there is no point in defining a disabled callback.
